server/src/main/java/org/opensearch/index/SegmentReplicationPressureService.java

```diff
@@ -154,4 +181,100 @@ public void setMaxAllowedStaleReplicas(double maxAllowedStaleReplicas) {
     public void setMaxReplicationTime(TimeValue maxReplicationTime) {
         this.maxReplicationTime = maxReplicationTime;
     }
+
+    @Override
+    public void close() throws IOException {
+        failStaleReplicaTask.close();
+    }
+
+    // Background Task to fail replica shards if they are too far behind primary shard.
+    final static class AsyncFailStaleReplicaTask extends AbstractAsyncTask {
+
+        final SegmentReplicationPressureService pressureService;
+
+        static final TimeValue INTERVAL = TimeValue.timeValueSeconds(30);
+
+        AsyncFailStaleReplicaTask(SegmentReplicationPressureService pressureService) {
+            super(logger, pressureService.threadPool, INTERVAL, true);
+            this.pressureService = pressureService;
+            rescheduleIfNecessary();
+        }
+
+        @Override
+        protected boolean mustReschedule() {
+            return true;
+        }
+
+        @Override
+        protected void runInternal() {
+            if (pressureService.isSegmentReplicationBackpressureEnabled) {
+                final SegmentReplicationStats stats = pressureService.tracker.getStats();
+                long highestCurrentReplicationTimeMillis = 0;
+                ShardId shardIdWithHighestCurrentReplicationTime = null;
+
+                // Find the shardId in node which is having stale replicas with highest current replication time.
+                // This way we only fail one shardId's stale replicas in every iteration of this background async task and there by decrease
+                // load gradually on node.
+                for (Map.Entry<ShardId, SegmentReplicationPerGroupStats> entry : stats.getShardStats().entrySet()) {
+                    final Set<SegmentReplicationShardStats> staleReplicas = pressureService.getStaleReplicas(
+                        entry.getValue().getReplicaStats()
+                    );
+                    for (SegmentReplicationShardStats staleReplica : staleReplicas) {
+                        if (staleReplica.getCurrentReplicationTimeMillis() > highestCurrentReplicationTimeMillis) {
```

Comment:
`highestCurrentReplicationTimeMillis` is never set to a value other than zero, is it? I think you can use the stream API, something like this:

```
stats.getShardStats().entrySet().stream()
    .flatMap(entry -> pressureService.getStaleReplicas(entry.getValue().getReplicaStats()).stream()
        .map(r -> Tuple.tuple(entry.getKey(), r.getCurrentReplicationTimeMillis())))
    .max(Comparator.comparingLong(Tuple::v2))
    .map(Tuple::v1);
    .ifPresent(shardId -> {
        ...
    });
```

Reply:
Sorry you are right, I missed reassigning `highestCurrentReplicationTimeMillis` in if condition. But Actually as you suggested using stream API makes this entire logic much cleaner. Thanks, I have updated the PR